File: docs/breez-sdk/snippets/swift/BreezSdkSnippets/Sources/Tokens.swift
import BigNumber
import BreezSdkSpark

func fetchTokenBalances(sdk: BreezSdk) async throws {
    // ANCHOR: fetch-token-balances
    // ensureSynced: true will ensure the SDK is synced with the Spark network
    // before returning the balance
    let info = try await sdk.getInfo(
        request: GetInfoRequest(
            ensureSynced: false
        ))

    // Token balances are a map of token identifier to balance
    let tokenBalances = info.tokenBalances
    for (tokenId, tokenBalance) in tokenBalances {
        print("Token ID: \(tokenId)")
        print("Balance: \(tokenBalance.balance)")
        print("Name: \(tokenBalance.tokenMetadata.name)")
        print("Ticker: \(tokenBalance.tokenMetadata.ticker)")
        print("Decimals: \(tokenBalance.tokenMetadata.decimals)")
    }
    // ANCHOR_END: fetch-token-balances
}

func fetchTokenMetadata(sdk: BreezSdk) async throws {
    // ANCHOR: fetch-token-metadata
    let response = try await sdk.getTokensMetadata(
        request: GetTokensMetadataRequest(tokenIdentifiers: [
            "<token identifier 1>", "<token identifier 2>",
        ]))

    let tokensMetadata = response.tokensMetadata
    for tokenMetadata in tokensMetadata {
        print("Token ID: \(tokenMetadata.identifier)")
        print("Name: \(tokenMetadata.name)")
        print("Ticker: \(tokenMetadata.ticker)")
        print("Decimals: \(tokenMetadata.decimals)")
        print("Max Supply: \(tokenMetadata.maxSupply)")
        print("Is Freezable: \(tokenMetadata.isFreezable)")
    }
    // ANCHOR_END: fetch-token-metadata
}

func receiveTokenPaymentSparkInvoice(sdk: BreezSdk) async throws -> ReceivePaymentResponse {
    // ANCHOR: receive-token-payment-spark-invoice
    let tokenIdentifier = "<token identifier>"
    let optionalDescription = "<invoice description>"
    let optionalAmount = BInt(5_000)
    // Optionally set the expiry UNIX timestamp in seconds
    let optionalExpiryTimeSeconds: UInt64 = 1_716_691_200
    let optionalSenderPublicKey = "<sender public key>"

    let response =
        try await sdk
        .receivePayment(
            request: ReceivePaymentRequest(
                paymentMethod: ReceivePaymentMethod.sparkInvoice(
                    amount: optionalAmount,
                    tokenIdentifier: tokenIdentifier,
                    expiryTime: optionalExpiryTimeSeconds,
                    description: optionalDescription,
                    senderPublicKey: optionalSenderPublicKey
                )
            ))

    let paymentRequest = response.paymentRequest
    print("Payment request: \(paymentRequest)")
    let receiveFeeSats = response.fee
    print("Fees: \(receiveFeeSats) token base units")
    // ANCHOR_END: receive-token-payment-spark-invoice

    return response
}

func sendTokenPayment(sdk: BreezSdk) async throws {
    // ANCHOR: send-token-payment
    let paymentRequest = "<spark address or invoice>"
    // Token identifier must match the invoice in case it specifies one.
    let tokenIdentifier = "<token identifier>"
    // Set the amount of tokens you wish to send. (requires 'import BigNumber')
    let optionalAmount = BInt(1_000)

    let prepareResponse = try await sdk.prepareSendPayment(
        request: PrepareSendPaymentRequest(
            paymentRequest: paymentRequest,
            amount: optionalAmount,
            tokenIdentifier: tokenIdentifier
        ))

    // If the fees are acceptable, continue to send the token payment
    if case let .sparkAddress(address, fee, tokenId) = prepareResponse.paymentMethod {
        print("Token ID: \(String(describing: tokenId))")
        print("Fees: \(fee) token base units")
    }
    if case let .sparkInvoice(invoice, fee, tokenId) = prepareResponse.paymentMethod {
        print("Token ID: \(String(describing: tokenId))")
        print("Fees: \(fee) token base units")
    }

    // Send the token payment
    let sendResponse = try await sdk.sendPayment(
        request: SendPaymentRequest(
            prepareResponse: prepareResponse,
            options: nil
        ))
    let payment = sendResponse.payment
    print("Payment: \(payment)")
    // ANCHOR_END: send-token-payment
}

func fetchConversionLimits(sdk: BreezSdk) async throws {
    // ANCHOR: fetch-conversion-limits
    // Fetch limits for converting Bitcoin to a token
    let fromBitcoinResponse = try await sdk.fetchConversionLimits(
        request: FetchConversionLimitsRequest(
            conversionType: ConversionType.fromBitcoin,
            tokenIdentifier: "<token identifier>"
        ))

    if let minFrom = fromBitcoinResponse.minFromAmount {
        print("Minimum BTC to convert: \(minFrom) sats")
    }
    if let minTo = fromBitcoinResponse.minToAmount {
        print("Minimum tokens to receive: \(minTo) base units")
    }

    // Fetch limits for converting a token to Bitcoin
    let toBitcoinResponse = try await sdk.fetchConversionLimits(
        request: FetchConversionLimitsRequest(
            conversionType: ConversionType.toBitcoin(
                fromTokenIdentifier: "<token identifier>"
            ),
            tokenIdentifier: nil
        ))

    if let minFrom = toBitcoinResponse.minFromAmount {
        print("Minimum tokens to convert: \(minFrom) base units")
    }
    if let minTo = toBitcoinResponse.minToAmount {
        print("Minimum BTC to receive: \(minTo) sats")
    }
    // ANCHOR_END: fetch-conversion-limits
}

func prepareSendPaymentTokenConversion(sdk: BreezSdk) async throws {
    // ANCHOR: prepare-send-payment-with-conversion
    let paymentRequest = "<spark address or invoice>"
    // Token identifier must match the invoice in case it specifies one.
    let tokenIdentifier = "<token identifier>"
    // Set the amount of tokens you wish to send. (requires 'import BigNumber')
    let optionalAmount = BInt(1_000)
    // Set to use Bitcoin funds to pay via conversion
    let optionalMaxSlippageBps = UInt32(50)
    let optionalCompletionTimeoutSecs = UInt32(30)
    let conversionOptions = ConversionOptions(
        conversionType: ConversionType.fromBitcoin,
        maxSlippageBps: optionalMaxSlippageBps,
        completionTimeoutSecs: optionalCompletionTimeoutSecs
    )

    let prepareResponse = try await sdk.prepareSendPayment(
        request: PrepareSendPaymentRequest(
            paymentRequest: paymentRequest,
            amount: optionalAmount,
            tokenIdentifier: tokenIdentifier,
            conversionOptions: conversionOptions
        ))

    // If the fees are acceptable, continue to send the token payment
    if let conversionEstimate = prepareResponse.conversionEstimate {
        print("Estimated conversion amount: \(conversionEstimate.amount) sats")
        print("Estimated conversion fee: \(conversionEstimate.fee) sats")
    }
    // ANCHOR_END: prepare-send-payment-with-conversion
}

File: docs/breez-sdk/snippets/kotlin_mpp_lib/shared/src/commonMain/kotlin/com/example/kotlinmpplib/Tokens.kt
package com.example.kotlinmpplib

import breez_sdk_spark.*
import com.ionspin.kotlin.bignum.integer.BigInteger

class Tokens {
    suspend fun fetchTokenBalances(sdk: BreezSdk) {
        // ANCHOR: fetch-token-balances
        try {
            // ensureSynced: true will ensure the SDK is synced with the Spark network
            // before returning the balance
            val info = sdk.getInfo(GetInfoRequest(false))

            // Token balances are a map of token identifier to balance
            val tokenBalances = info.tokenBalances
            for ((tokenId, tokenBalance) in tokenBalances) {
                println("Token ID: $tokenId")
                println("Balance: ${tokenBalance.balance}")
                println("Name: ${tokenBalance.tokenMetadata.name}")
                println("Ticker: ${tokenBalance.tokenMetadata.ticker}")
                println("Decimals: ${tokenBalance.tokenMetadata.decimals}")
            }
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: fetch-token-balances
    }

    suspend fun fetchTokenMetadata(sdk: BreezSdk) {
        // ANCHOR: fetch-token-metadata
        try {
            val response = 
                sdk.getTokensMetadata(
                    GetTokensMetadataRequest(
                        tokenIdentifiers = listOf("<token identifier 1>", "<token identifier 2>")
                )
            )   

            val tokensMetadata = response.tokensMetadata
            for (tokenMetadata in tokensMetadata) {
                println("Token ID: ${tokenMetadata.identifier}")
                println("Name: ${tokenMetadata.name}")
                println("Ticker: ${tokenMetadata.ticker}")
                println("Decimals: ${tokenMetadata.decimals}")
                println("Max Supply: ${tokenMetadata.maxSupply}")
                println("Is Freezable: ${tokenMetadata.isFreezable}")
            }
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: fetch-token-metadata
    }

    suspend fun receiveTokenPaymentSparkInvoice(sdk: BreezSdk) {
        // ANCHOR: receive-token-payment-spark-invoice
        try {
            val tokenIdentifier = "<token identifier>"
            val optionalDescription = "<invoice description>"
            // Kotlin MPP (BigInteger from com.ionspin.kotlin.bignum.integer, which is included in
            // package)
            val optionalAmount = BigInteger.fromLong(5_000L)
            // Android (BigInteger from java.math)
            // val optionalAmount = BigInteger.valueOf(5_000L)
            // Optionally set the expiry UNIX timestamp in seconds
            val optionalExpiryTimeSeconds = 1716691200.toULong()
            val optionalSenderPublicKey = "<sender public key>"

            val request = ReceivePaymentRequest(
                ReceivePaymentMethod.SparkInvoice(
                    tokenIdentifier = tokenIdentifier,
                    description = optionalDescription,
                    amount = optionalAmount,
                    expiryTime = optionalExpiryTimeSeconds,
                    senderPublicKey = optionalSenderPublicKey
                )
            )
            val response = sdk.receivePayment(request)

            val paymentRequest = response.paymentRequest
            println("Payment request: $paymentRequest")
            val receiveFee = response.fee
            println("Fees: $receiveFee token base units")
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: receive-token-payment-spark-invoice
    }

    suspend fun sendTokenPayment(sdk: BreezSdk) {
        // ANCHOR: send-token-payment
        try {
            val paymentRequest = "<spark address or invoice>"
            // Token identifier must match the invoice in case it specifies one.
            val tokenIdentifier = "<token identifier>"
            // Set the amount of tokens you wish to send.
            // Kotlin MPP (BigInteger from com.ionspin.kotlin.bignum.integer, which is included in
            // package)
            val amount = BigInteger.fromLong(1_000L)
            // Android (BigInteger from java.math)
            // val amount = BigInteger.valueOf(1_000L)
            val optionalPayAmount = PayAmount.Token(amount = amount, tokenIdentifier = tokenIdentifier)

            val prepareResponse =
                sdk.prepareSendPayment(
                    PrepareSendPaymentRequest(
                        paymentRequest = paymentRequest,
                        payAmount = optionalPayAmount,
                    )
                )

            // If the fees are acceptable, continue to send the token payment
            when (val method = prepareResponse.paymentMethod) {
                is SendPaymentMethod.SparkAddress -> {
                    println("Token ID: ${method.tokenIdentifier}")
                    println("Fees: ${method.fee} token base units")
                }
                is SendPaymentMethod.SparkInvoice -> {
                    println("Token ID: ${method.tokenIdentifier}")
                    println("Fees: ${method.fee} token base units")
                }
                else -> {}
            }

            // Send the token payment
            val sendResponse =
                sdk.sendPayment(
                    SendPaymentRequest(prepareResponse = prepareResponse, options = null)
                )
            val payment = sendResponse.payment
            println("Payment: $payment")
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: send-token-payment
    }

    suspend fun fetchConversionLimits(sdk: BreezSdk) {
        // ANCHOR: fetch-conversion-limits
        try {
            // Fetch limits for converting Bitcoin to a token
            val fromBitcoinResponse = sdk.fetchConversionLimits(
                FetchConversionLimitsRequest(
                    conversionType = ConversionType.FromBitcoin,
                    tokenIdentifier = "<token identifier>"
                )
            )

            if (fromBitcoinResponse.minFromAmount != null) {
                println("Minimum BTC to convert: ${fromBitcoinResponse.minFromAmount} sats")
            }
            if (fromBitcoinResponse.minToAmount != null) {
                println("Minimum tokens to receive: ${fromBitcoinResponse.minToAmount} base units")
            }

            // Fetch limits for converting a token to Bitcoin
            val toBitcoinResponse = sdk.fetchConversionLimits(
                FetchConversionLimitsRequest(
                    conversionType = ConversionType.ToBitcoin(
                        fromTokenIdentifier = "<token identifier>"
                    ),
                    tokenIdentifier = null
                )
            )

            if (toBitcoinResponse.minFromAmount != null) {
                println("Minimum tokens to convert: ${toBitcoinResponse.minFromAmount} base units")
            }
            if (toBitcoinResponse.minToAmount != null) {
                println("Minimum BTC to receive: ${toBitcoinResponse.minToAmount} sats")
            }
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: fetch-conversion-limits
    }

    suspend fun prepareSendPaymentTokenConversion(sdk: BreezSdk) {
        // ANCHOR: prepare-send-payment-with-conversion
        try {
            val paymentRequest = "<spark address or invoice>"
            // Token identifier must match the invoice in case it specifies one.
            val tokenIdentifier = "<token identifier>"
            // Set the amount of tokens you wish to send.
            // Kotlin MPP (BigInteger from com.ionspin.kotlin.bignum.integer, which is included in
            // package)
            val amount = BigInteger.fromLong(1_000L)
            // Android (BigInteger from java.math)
            // val amount = BigInteger.valueOf(1_000L)
            val optionalPayAmount = PayAmount.Token(amount = amount, tokenIdentifier = tokenIdentifier)
            // set to use Bitcoin funds to pay via conversion
            val optionalMaxSlippageBps = 50u
            val optionalCompletionTimeoutSecs = 30u
            val conversionOptions = ConversionOptions(
                conversionType = ConversionType.FromBitcoin,
                maxSlippageBps = optionalMaxSlippageBps,
                completionTimeoutSecs = optionalCompletionTimeoutSecs
            )

            val prepareResponse =
                sdk.prepareSendPayment(
                    PrepareSendPaymentRequest(
                        paymentRequest = paymentRequest,
                        payAmount = optionalPayAmount,
                        conversionOptions = conversionOptions,
                    )
                )

            // If the fees are acceptable, continue to send the token payment
            prepareResponse.conversionEstimate?.let { conversionEstimate ->
                println("Estimated conversion amount: ${conversionEstimate.amount} sats")
                println("Estimated conversion fee: ${conversionEstimate.fee} sats")
            }
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: prepare-send-payment-with-conversion
    }
}
